packages/jest-config/src/utils.js

```diff
@@ -102,43 +102,135 @@ export const _replaceRootDirTags = (rootDir: string, config: any) => {
   return config;
 };
 
-/**
- * Finds the test environment to use:
- *
- * 1. looks for jest-environment-<name> relative to project.
- * 1. looks for jest-environment-<name> relative to Jest.
- * 1. looks for <name> relative to project.
- * 1. looks for <name> relative to Jest.
- */
-export const getTestEnvironment = (config: Object) => {
-  const env = replaceRootDirInPath(config.rootDir, config.testEnvironment);
-  let module = Resolver.findNodeModule(`jest-environment-${env}`, {
-    basedir: config.rootDir,
+export const resolveWithPrefix = (
+  resolver: ?string,
+  {
+    rootDir,
+    filePath,
+    prefix,
+  }: {
+    rootDir: string,
+    filePath: string,
+    prefix: string,
+  },
+) => {
+  const fileName = replaceRootDirInPath(rootDir, filePath);
+  let module = Resolver.findNodeModule(`${prefix}${fileName}`, {
+    basedir: rootDir,
+    resolver,
   });
   if (module) {
     return module;
   }
 
   try {
-    return require.resolve(`jest-environment-${env}`);
+    return require.resolve(`${prefix}${fileName}`);
   } catch (e) {}
 
-  module = Resolver.findNodeModule(env, {basedir: config.rootDir});
+  module = Resolver.findNodeModule(fileName, {
+    basedir: rootDir,
+    resolver,
+  });
   if (module) {
     return module;
   }
 
   try {
-    return require.resolve(env);
+    return require.resolve(fileName);
   } catch (e) {}
+};
 
-  throw createValidationError(
-    `  Test environment ${chalk.bold(
-      env,
-    )} cannot be found. Make sure the ${chalk.bold(
-      'testEnvironment',
-    )} configuration option points to an existing node module.`,
-  );
+/**
+ * Finds the test environment to use:
+ *
+ * 1. looks for jest-environment-<name> relative to project.
+ * 1. looks for jest-environment-<name> relative to Jest.
+ * 1. looks for <name> relative to project.
+ * 1. looks for <name> relative to Jest.
+ */
+export const getTestEnvironment = ({
+  rootDir,
+  testEnvironment: filePath,
+}: Object) => {
+  const module = resolveWithPrefix(null, {
+    filePath,
+    prefix: 'jest-environment-',
+    rootDir,
+  });
+
+  if (!module) {
+    throw createValidationError(
```

Comment:
seems like you could throw this in `resolveWithPrefix`? They're all identical except for, in this case, `Test environment` and `'testEnvironment'`.

So you could end up with e.g.

```js
export const getTestEnvironment = ({
  rootDir,
  testEnvironment: filePath,
}: Object) => {
  return resolveWithPrefix(null, {
    filePath,
    prefix: 'jest-environment-',
    rootDir,
    optionName: 'testEnvironment',
    humanOptionName: 'Test environment',
  });
};
```

or something like that

Reply:
I originally left it out because the watch plugin validation happens as an element in an array. But I guess it is fine to show the same message.